docs/docs/guides/03-typescript.md

```diff
@@ -2,4 +2,226 @@
 title: Typescript
 ---
 
-🚧 coming soon 🚧
+# Development Guidelines
+
+This document provides guidelines for developing on the Strapi codebase using TypeScript. It covers various topics, including types location, generics, breaking changes, factories and more.
+
+## 1. Type Location
+
+When working with types on the Strapi codebase, it is essential to consider the location of types. These are the three common approaches and situations we document:
+
+### a. Colocation
+
+In the colocation approach, type definitions are colocated with the corresponding code. This means that types are defined in the same file or module where they are used.
+
+**This approach is suitable when the types are specific to a particular module**.
+
+_Note: In some situations, colocated types can still be exported if needed by other modules_
+
+Example:
+
+```typescript
+// config.ts
+interface Config {
+  prop1: string;
+  prop2: number;
+}
+
+function update(config: Config): Config {
+  // ...
+}
+```
+
+### b. Definitions Files
```

Comment:
Now that I think more about this (and taken some time to look at other libraries), I'm wondering whether having regular .ts file for type declaration wouldn't make more sense?
And then keeping .d.ts for module declaration & extension.

My only concern is that it adds some level of complexity + you'll sometimes end up with .d.ts next to .ts which might not be very clear.

What do you think @joshuaellis / @innerdvations / @alexandrebodin ?

Reply:
In theory I think you're right. We should be using regular .ts files and then exporting whatever we explicitly want to export in each module in a .d.ts file. Pretty close to what we're already doing but renaming most of our .d.ts -> .ts

However, I don't think it hurts anything to keep them in .d.ts files, especially since we still have so much javascript and jsdoc usage in the codebase, and they tend to be picked up better by editors than regular .ts files. Plus .d.ts enforces that we ONLY include typings in those files and no code can accidentally slip in (although maybe there's a rule for that we could add for the types dir? no idea)

I don't have a strong opinion either way, curious to hear arguments from others though.

Reply:
someone did a twitter thread on this – https://twitter.com/mattpocockuk/status/1672918127414333440 https://twitter.com/mattpocockuk/status/1673017324587171842

Reply:
Yes I saw this one, but I don't think it fits very well Strapi's context.
I mean, it basically ends up being the pros/cons that we've listed above :confused:

Reply:
I'm leaning towards `.ts` at this point for the skipLibCheck, and that it seems to be best practices (after reading through the twitter debate and links there). I don't see any major pros to `.d.ts`